pysnark/runtime.py

```diff
@@ -483,99 +546,169 @@ def __trunc__(self): return NotImplemented
         
         # Trick from Pinocchio paper: if self is zero then ret=1 by first eq,
         # if self is nonzero then ret=0 by second eq
-        add_constraint_unsafe(self, wit, LinComb.ONE_SAFE-ret)
+        add_constraint_unsafe(self, wit, LinComb.ONE_SAFE - ret)
         add_constraint_unsafe(self, ret, LinComb.ZERO)
-        
-        return ret
+
+        return LinCombBool(ret)
+
+    def check_nonzero(self):
+        """
+        Checks whether a LinComb is nonzero
+        Costs 4 constraints
+        """
+        from pysnark.boolean import LinCombBool
+
+        ret = PrivVal(1 if self.value != 0 else 0)
+        wit = PrivVal(backend.fieldinverse(self.value + (self.value == 0)))  # Add self.value == 0 to prevent ZeroDivisionError
+
+        # Use nonzero check gadget from Bulletproofs
+        add_constraint_unsafe(self, 1 - ret, LinComb.ZERO)
+        add_constraint_unsafe(self, wit, ret)
+
+        return LinCombBool(ret)
     
     def assert_zero(self, err=None):
+        """
+        Ensures a LinComb is zero
+        Costs 1 constraint
+        """
         if (not ignore_errors()) and self.value!=0:
-            raise ValueError(err if err is not None else "value " + str(self.value) + " is not zero")
+            raise AssertionError(err if err is not None else str(self.value) + " is not zero")
             
-        add_constraint(LinComb.ZERO, LinComb.ZERO, self)        
+        add_constraint(LinComb.ZERO, LinComb.ZERO, self)
             
     def assert_nonzero(self, err=None):
+        """
+        Ensures a LinComb is nonzero
+        Costs 1 constraint
+        """
         if (is_guard()) and self.value!=0:
             wit = PrivVal(backend.fieldinverse(self.value))    
         elif ignore_errors():
             wit = PrivVal(0)
         else:
-            raise ValueError(err if err is not None else "value " + str(self.value) + " is not zero")
+            raise AssertionError(err if err is not None else str(self.value) + " is zero")
         
         add_constraint(self, wit, LinComb.ONE, check=False)
-        
-    def if_else(self, ifval, elseval):
-        from .branching import if_then_else
-        return if_then_else(self, ifval, elseval)
-    
+
+    def assert_range(self, rangemin, rangemax, err=None):
+        """
+        Ensures a LinComb is within the range [rangemin, rangemax]
+        Costs 4 * bitlength + 3 constraints
+        """
+        rangemin = LinComb._ensurelc(rangemin)
+        rangemax = LinComb._ensurelc(rangemax)
+
+        if not ignore_errors():
+            if self.value < rangemin.value or self.value >= rangemax.value:
+                raise AssertionError(err if err is not None else str(self.value) + " is not in the range [" + str(rangemin.value) + "," + str(rangemax.value) + "]")
+
+        # Use bounds check gadget from Bulletproofs
+        a = self - rangemin
+        b = rangemax - self
+
+        (a + b).assert_eq(rangemax - rangemin, err)
```

Comment:
It seems to me that the constraint on (a+b) is not needed since this will be true by construction?

Reply:
I'm not too sure. I stole this from here https://github.com/lovesh/bulletproofs-r1cs-gadgets/blob/master/src/gadget_range_proof.rs#L167

Reply:
I think their setting is different. In our case, doing a=self-rangemin will just set a to be copy of the lists of variables in self plus a copy of the list of variables of rangemin, with their coefficnets inverted. Same for b. Doing a+b will just concatenate their lists of variables, which will cancel out the variables of self since they occur with positive sign in a and with negative sign in b. So you end up with a linear combination that is by construction the same as rangemax-rangemin. In the bulletproofs case there is probably a good reason why in their setting the constraint is needed, but in our case it souldn't be.

Reply:
Fixed!